src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,388 @@
+
+        /// <summary>
+        /// The text snapshot from when the document symbol request was made.
+        /// </summary>
+        private ITextSnapshot? LspSnapshot { get; set; }
+
+        /// <summary>
+        /// Keeps track of the current primary and secondary text views.
+        /// </summary>
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
+
+        /// <summary>
+        /// Stores the latest document model returned by GetModelAsync to be used by UpdateModelAsync.
+        /// </summary>
+        private ImmutableArray<DocumentSymbolViewModel> LatestFetchedDocumentModel { get; set; }
+
+        /// <summary>
+        /// Is true when LatestFetchedDocumentModel is not empty.
+        /// </summary>
+        [MemberNotNullWhen(true, nameof(LspSnapshot))]
+        private bool LatestFetchedDocumentModelIsNotEmpty { get; set; }
+
+        public DocumentOutlineControl(
+            ILanguageServiceBrokerShim languageServiceBroker,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            InitializeComponent();
+
+            ThreadingContext = threadingContext;
+            LanguageServiceBroker = languageServiceBroker;
+            EditorAdaptersFactoryService = editorAdaptersFactoryService;
+            CodeWindow = codeWindow;
+            ComEventSink.Advise<IVsCodeWindowEvents>(codeWindow, this);
+            SortOption = SortOption.Order;
+
+            _getModelQueue = new AsyncBatchingWorkQueue(
+                    DelayTimeSpan.Short,
+                    GetModelAsync,
+                    asyncListener,
+                    threadingContext.DisposalToken);
+
+            _updateModelQueue = new AsyncBatchingWorkQueue(
+                    DelayTimeSpan.NearImmediate,
+                    UpdateModelAsync,
+                    asyncListener,
+                    threadingContext.DisposalToken);
+
+            _highlightNodeQueue = new AsyncBatchingWorkQueue(
+                    DelayTimeSpan.NearImmediate,
+                    HightlightNodeAsync,
+                    asyncListener,
+                    threadingContext.DisposalToken);
+
+            codeWindow.GetPrimaryView(out var pTextViewPrimary);
+            StartTrackingView(pTextViewPrimary, out var textViewPrimary);
+
+            // Primary text view should always exist on window initialization unless an error is thrown.
+            if (textViewPrimary is null)
+                return;
+
+            codeWindow.GetSecondaryView(out var pTextViewSecondary);
+            if (pTextViewSecondary is not null)
+                StartTrackingView(pTextViewSecondary, out var _);
+
+            StartGetModelTask();
+        }
+
+        int IVsCodeWindowEvents.OnNewView(IVsTextView pView)
+        {
```

Comment:
for teh IVs entrypoints, can you add a threading assert that you're on hte UI thread?